src/components/SmartModeInput/index.tsx

```diff
@@ -0,0 +1,198 @@
+
+  useEffect(() => {
+    if (to && isAddress(to)) {
+      fetchABI(
+        proxyImplementationAddress && isAddress(proxyImplementationAddress)
+          ? (proxyImplementationAddress as Address)
+          : to
+      );
+    } else {
+      setAbi(null);
+      setSelectedFunction(null);
+      setFunctionInputs({});
+    }
+  }, [to, selectedChain, proxyImplementationAddress]);
+
+  const fetchABI = async (address: Address) => {
+    setIsLoading(true);
+    const apiKey =
+      selectedChain === 'optimism-sepolia'
+        ? process.env.NEXT_PUBLIC_OPTIMISM_ETHERSCAN_API_KEY
+        : process.env.NEXT_PUBLIC_ARBITRUM_ETHERSCAN_API_KEY;
+
+    const apiUrl =
+      selectedChain === 'optimism-sepolia'
+        ? 'https://api-sepolia-optimistic.etherscan.io/api'
+        : 'https://api-sepolia.arbiscan.io/api';
+
+    try {
+      const response = await fetch(
+        `${apiUrl}?module=contract&action=getabi&address=${address}&apikey=${apiKey}`
+      );
+      const data = await response.json();
+      if (data.status === '1') {
+        setAbi(JSON.parse(data.result));
+      } else {
+        toast.error('Failed to fetch ABI');
+      }
+    } catch (error) {
+      console.error('Error fetching ABI:', error);
+      toast.error('Error fetching ABI');
+    } finally {
+      setIsLoading(false);
+    }
+  };
+
+  const handleFunctionSelect = (functionName: string) => {
+    setSelectedFunction(functionName);
+    setFunctionInputs({});
+  };
+
+  const handleInputChange = (name: string, value: string) => {
+    setFunctionInputs((prev) => ({ ...prev, [name]: value }));
+  };
+
+  const generateData = () => {
+    if (!abi || !selectedFunction) return;
+
+    const functionAbi = abi.find((item: any) => item.name === selectedFunction);
+    if (!functionAbi) return;
+
+    try {
+      const data = encodeFunctionData({
+        abi: [functionAbi],
+        functionName: selectedFunction,
+        args: functionAbi.inputs.map(
+          (input: any) => functionInputs[input.name]
+        ),
+      });
+      onDataGenerated(data);
+    } catch (error) {
+      console.error('Error encoding function data:', error);
+      toast.error('Invalid input data');
+    }
+  };
+
+  const writableFunctions =
+    abi?.filter(
+      (item: any) =>
+        item.type === 'function' && !item.stateMutability.includes('view')
+    ) || [];
+
+  return (
+    <div className="space-y-4">
+      {isLoading && <p className="text-sm text-gray-500">Loading ABI...</p>}
+
+      <Input label="To Address" value={to} isReadOnly size="sm" isDisabled />
+      <Tooltip content="Use this if the contract is a proxy. Enter the implementation contract address.">
+        <Input
+          label="Proxy Implementation Address (optional)"
+          placeholder="0x"
+          type="text"
+          value={proxyImplementationAddress}
+          onChange={(e) => setProxyImplementationAddress(e.target.value)}
+          size="sm"
+        />
+      </Tooltip>
```

Comment:
I think we can leverage etherscan's api to automatically pulled the implementation address of a verified proxy contract on etherscan: https://ethereum.stackexchange.com/questions/135333/is-it-possible-to-pull-proxy-contracts-through-an-ethereum-based-api

This is the result queried with `https://api-sepolia-optimistic.etherscan.io/api?module=contract&action=getsourcecode&address=0x5fd84259d66Cd46123540766Be93DFE6D43130D7`:
```
{
  "status": "1",
  "message": "OK",
  "result": [
    {
      "SourceCode": ...
      ...
      "Implementation": "0x101e1c9757c99867a39ceb41117b44f2292cb253",
      ...
    }
  ]
}
```

Reply:
And if the proxy is not verified on etherscan, then we provide this additional manual input tool in case user needs it.